FPC LRU cache related changes

mmanoj opened the pull request:
Please sign (check) the below before submitting the Pull Request:

- [x] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [x] I have read the contributing guide lines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [x] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Link to the related [issue](https://github.com/ntop/nDPI/issues):


Describe changes:
changes related to https://github.com/ntop/nDPI/issues/2322
item# 2
2.1.Create a new LRU cache where the key is the pair "Client_IP - Resolved_IP" and the value is a protocol id:
2.2 For each flow (at the very first packet) lookup into the cache using "SRC IP - DST IP" as key and, if a match is found, save the protocol id into ndpi_flow_struct structure as a "fpc result"

mmanoj committed:
Add files via upload

update fpc lru cache related changes

mmanoj committed:
Add files via upload

fpc lru cache related changes

IvanNardi commented on src/include/ndpi_private.h:
```diff
@@ -168,6 +168,9 @@ struct ndpi_global_context {
   struct ndpi_lru_cache *msteams_global_cache;
+  
+  /* NDPI_FIRST_PKT_CLASSIFICATION_CACHE */
+  struct ndpi_lru_cache *fpc_global_cache;
```
Since we are using this cache for DNS mapping, I would prefer `fpc_dns_global_cache` name.
Same change pretty much everywhere....

mmanoj replied:
@IvanNardi 
Thank you very much for your kind review and feedback.i'm working on the changes and update you once tested in my local env.

IvanNardi commented on src/include/ndpi_typedefs.h:
```diff
@@ -999,7 +1000,7 @@ typedef enum {
   NDPI_FPC_CONFIDENCE_IP,                       /* FPC based on IP address */
   NDPI_FPC_CONFIDENCE_DNS,                      /* FPC based on DNS information */
-
+  NDPI_FPC_CONFIDENCE_CACHE,
```
Remove it. We already have `NDPI_FPC_CONFIDENCE_DNS`

IvanNardi commented on src/include/ndpi_private.h:
```diff
@@ -220,7 +223,10 @@ struct ndpi_detection_module_config_struct {
-
+  int fpc_cache_num_entries;
+  int fpc_cache_ttl;
+  int fpc_cache_scope;
```
name

IvanNardi commented on src/include/ndpi_private.h:
```diff
@@ -361,6 +367,9 @@ struct ndpi_detection_module_struct {
   struct ndpi_lru_cache *msteams_cache;
+  
+  /* NDPI_FIRST_PKT_CLASSIFICATION_CACHE */
+  struct ndpi_lru_cache *fpc_cache;
```
name

IvanNardi commented on src/include/ndpi_typedefs.h:
```diff
@@ -747,6 +747,7 @@ typedef enum {
   NDPI_LRUCACHE_TLS_CERT,
   NDPI_LRUCACHE_MINING,
   NDPI_LRUCACHE_MSTEAMS,
+  NDPI_LRUCACHE_FPC, /* First Pkt Classification */
```
name... and all the code below

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -9421,6 +9451,9 @@ const char *ndpi_fpc_confidence_get_name(ndpi_fpc_confidence_t fpc_confidence)
     return "DNS";
+    
+  case NDPI_FPC_CONFIDENCE_CACHE:
+      return "FPC cache";
```
Remove it

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -11293,8 +11330,11 @@ static const struct cfg_param {
-
-
+//ToDo:add fpc
+{ NULL,            "lru.fpc.size",                        "1024", "0", "16777215", CFG_PARAM_INT, __OFF(msteams_cache_num_entries), NULL },
```
indentation and, more important, the right names!!! `lru.fpc_dns.size` and `fpc_dns_cache_num_entries`
Same the two lines below

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -8375,6 +8376,33 @@ static ndpi_protocol ndpi_internal_detection_process_packet(struct ndpi_detectio
+/* For each flow (at the very first packet) lookup into the cache using "SRC IP - DST IP" as key and, if a match is found, save the protocol id into ndpi_flow_struct structure as a "fpc result" */
+//printf("\nApp proto from ret-->%u\n",ret.app_protocol);
+if(!ndpi_str->fpc_cache){
+ printf("\n No FPC cache..\n");
```
For logging please use `NDPI_LOG_DBG` and similar

IvanNardi reviewed (changes requested):
@mmanoj, some things to improve; this is only a preliminary review.
The most obvious missing piece is the cache initialization! You must add it in `ndpi_finalize_initialization`
Please, be sure that the code compile (at least on your machine), before pushing it

mmanoj committed:
Update for review changes

Update for review changes

mmanoj committed:
Update as per review comments

Update as per review comments

mmanoj reviewed (commented):
Update as per review comments.Howeer need some clarifications regarding section of "first packet of this flow to be analyzed" in ndpi_main.c we can discuss it after this review.

mmanoj commented:
> @mmanoj, some things to improve; this is only a preliminary review. The most obvious missing piece is the cache initialization! You must add it in `ndpi_finalize_initialization` Please, be sure that the code compile (at least on your machine), before pushing it

All changes done and compiled in local env with latest dev branch.

IvanNardi commented on src/include/ndpi_private.h:
```diff
@@ -168,6 +168,9 @@ struct ndpi_global_context {
   /* NDPI_PROTOCOL_MSTEAMS */
   struct ndpi_lru_cache *msteams_global_cache;
+  
+  /* NDPI_FIRST_PKT_CLASSIFICATION_CACHE */
```
/* FPC DNS cache */

IvanNardi commented on src/include/ndpi_private.h:
```diff
@@ -361,6 +367,9 @@ struct ndpi_detection_module_struct {
   /* NDPI_PROTOCOL_MSTEAMS */
   struct ndpi_lru_cache *msteams_cache;
+  
+  /* NDPI_FIRST_PKT_CLASSIFICATION_CACHE */
```
As above

IvanNardi commented on src/include/ndpi_typedefs.h:
```diff
@@ -747,6 +747,7 @@ typedef enum {
   NDPI_LRUCACHE_TLS_CERT,
   NDPI_LRUCACHE_MINING,
   NDPI_LRUCACHE_MSTEAMS,
+  NDPI_LRUCACHE_FPC_DNS, /* First Pkt Classification */
```
As above

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -3829,6 +3831,23 @@ int ndpi_finalize_initialization(struct ndpi_detection_module_struct *ndpi_str)
+  if(ndpi_str->cfg.fpc_dns_cache_num_entries > 0) {
+    if(ndpi_str->cfg.fpc_dns_cache_scope == NDPI_LRUCACHE_SCOPE_GLOBAL) {
+      if(!ndpi_str->g_ctx->fpc_dns_global_cache) {
+        ndpi_str->g_ctx->fpc_dns_global_cache = ndpi_lru_cache_init(ndpi_str->cfg.fpc_dns_cache_num_entries,                                                                   ndpi_str->cfg.fpc_dns_cache_ttl, 1);
```
the indentation seems wrong....

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -4166,6 +4185,10 @@ void ndpi_exit_detection_module(struct ndpi_detection_module_struct *ndpi_str) {
     if(!ndpi_str->cfg.msteams_cache_scope &&
        ndpi_str->msteams_cache)
       ndpi_lru_free_cache(ndpi_str->msteams_cache);
+    //ToDo:add fpc details
```
I think we can safely remove this comment

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -8448,6 +8526,8 @@ static ndpi_protocol ndpi_internal_detection_process_packet(struct ndpi_detectio
 
     fpc_check_ip(ndpi_str, flow);
+    /* Update fpc cache */
+    ndpi_fpc_cache_update(ndpi_str,flow);
```
Here we need to lookup into the cache, not update it!

mmanoj replied:
I will arrange the changes, are we going to update the cache in DNS dissector? or any other places? Please advice.

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -10250,6 +10330,10 @@ int ndpi_get_lru_cache_stats(struct ndpi_global_context *g_ctx,
   case NDPI_LRUCACHE_MSTEAMS:
     ndpi_lru_get_stats(is_local ? ndpi_struct->msteams_cache : g_ctx->msteams_global_cache, stats);
     return 0;
+    //ToDo:add fpc cache details
```
Remove the comment, please

IvanNardi reviewed (changes requested):
The general idea of the cache is:
* add an entry in the cache with DNS traffic. In the DNS code, if `ndpi_match_host_subprotocol()` return a protocol different than `NDPI_PROTOCOL_UNKNOWN`, we should do somethig like `ndpi_lru_add_to_cache(ndpi_str->fpc_dns_cache,  key, ret.app_protocol, ndpi_get_current_time(flow));`.
The key is something like SRC_FLOW_IP + ip_required_in_the_dns_query

* perform a lookup in `ndpi_internal_detection_process_packet` just after `fpc_check_ip()` call. If there is an entry, call something like `fpc_update(ndpi_str, flow, NDPI_PROTOCOL_UNKNOWN, protocol_returned_from_the_cache, NDPI_FPC_CONFIDENCE_DNS)`

Any specific doubts?

IvanNardi commented on src/lib/ndpi_main.c:
```diff
@@ -8375,6 +8428,31 @@ static ndpi_protocol ndpi_internal_detection_process_packet(struct ndpi_detectio
+			 &fpc_dns_cached_proto, 0 /* Don't remove it as it can be used for other connections */,
+			 ndpi_get_current_time(flow))) {
+    NDPI_LOG_DBG(ndpi_str,"\nFound from FPC cache...%u\n",fpc_dns_cached_proto);
+    ndpi_set_detected_protocol(ndpi_str, flow, fpc_dns_cached_proto, NDPI_PROTOCOL_UNKNOWN, NDPI_FPC_CONFIDENCE_DNS);
```
The goal of this work is to set the FPC classification, not the "general" one! So no calls to `ndpi_set_detected_protocol()`!!

mmanoj replied:
I will arrange fixes as advice.

mmanoj commented:
> The general idea of the cache is:
> 
>     * add an entry in the cache with DNS traffic. In the DNS code, if `ndpi_match_host_subprotocol()` return a protocol different than `NDPI_PROTOCOL_UNKNOWN`, we should do somethig like `ndpi_lru_add_to_cache(ndpi_str->fpc_dns_cache,  key, ret.app_protocol, ndpi_get_current_time(flow));`.
>       The key is something like SRC_FLOW_IP + ip_required_in_the_dns_query
> 
>     * perform a lookup in `ndpi_internal_detection_process_packet` just after `fpc_check_ip()` call. If there is an entry, call something like `fpc_update(ndpi_str, flow, NDPI_PROTOCOL_UNKNOWN, protocol_returned_from_the_cache, NDPI_FPC_CONFIDENCE_DNS)`
> 
> 
> Any specific doubts?

can you explain bit more about this? SRC_FLOW_IP + ip_required_in_the_dns_query
 is this flow source IP and DNS response IP (like A or AAA address?)

IvanNardi commented:
> 
> can you explain bit more about this? SRC_FLOW_IP + ip_required_in_the_dns_query is this flow source IP and DNS response IP (like A or AAA address?)

That was the idea, but I was wrong (because of ipv4 and ipv6 possible mismatch) We should use only one address as key.

Anyway, let try together to analyze what we should do step by step.
In this simple trace there is a DNS flow and a TCP flow: [fpc_dns.zip](https://github.com/user-attachments/files/16239123/fpc_dns.zip)

1) We should take a look at every DNS A or AAAA replies. If the domain required provides a sub-classification, i.e. if `ndpi_match_host_subprotocol()` in DNS code returns a protocol different than `NDPI_PROTOCOL_UNKNOWN`, we should save that information into the cache. I think we can always add only one entry, even if there are multiple answers in the DNS reply
   *) Packet 2 is a DNS AAAA reply with Youtube as subclassification -> add an entry with key "2a00:1450:4002:416::200e" (i.e. the first answer) and protocol `NDPI_PROTOCOL_YOUTUBE`

2) At the very first pkt of every flow, we should lookup into the cache using the server address (i.e. `flow->s_addr`), in `ndpi_internal_detection_process_packet` just after `fpc_check_ip()` call. If a match is found, we should call `fpc_update(ndpi_str, flow, NDPI_PROTOCOL_UNKNOWN, protocol_returned_from_the_cache, NDPI_FPC_CONFIDENCE_DNS)`
   *) At packet 3 we should lookup using "2a00:1450:4002:416::200e"  as key (the server address). The returned protocol is `NDPI_PROTOCOL_YOUTUBE`, so the FPC of this flow is Youtube with confidence "dns"

If everything is implemented correctly we should got something like:
```
./example/ndpiReader -t -i ~/fpc_dns.pcapng -v2


	1	TCP [2001:b07:a3d:c112:9326:58a1:68f7:5712]:40088 <-> [2a00:1450:4002:410::200e]:443 ... [FPC: 124/Youtube, Confidence: DNS]
```

Am I missing something?

mmanoj commented:
> > can you explain bit more about this? SRC_FLOW_IP + ip_required_in_the_dns_query is this flow source IP and DNS response IP (like A or AAA address?)
> 
> That was the idea, but I was wrong (because of ipv4 and ipv6 possible mismatch) We should use only one address as key.
> 
> Anyway, let try together to analyze what we should do step by step. In this simple trace there is a DNS flow and a TCP flow: [fpc_dns.zip](https://github.com/user-attachments/files/16239123/fpc_dns.zip)
> 
>     1. We should take a look at every DNS A or AAAA replies. If the domain required provides a sub-classification, i.e. if `ndpi_match_host_subprotocol()` in DNS code returns a protocol different than `NDPI_PROTOCOL_UNKNOWN`, we should save that information into the cache. I think we can always add only one entry, even if there are multiple answers in the DNS reply
>        *) Packet 2 is a DNS AAAA reply with Youtube as subclassification -> add an entry with key "2a00:1450:4002:416::200e" (i.e. the first answer) and protocol `NDPI_PROTOCOL_YOUTUBE`
> 
>     2. At the very first pkt of every flow, we should lookup into the cache using the server address (i.e. `flow->s_addr`), in `ndpi_internal_detection_process_packet` just after `fpc_check_ip()` call. If a match is found, we should call `fpc_update(ndpi_str, flow, NDPI_PROTOCOL_UNKNOWN, protocol_returned_from_the_cache, NDPI_FPC_CONFIDENCE_DNS)`
>        *) At packet 3 we should lookup using "2a00:1450:4002:416::200e"  as key (the server address). The returned protocol is `NDPI_PROTOCOL_YOUTUBE`, so the FPC of this flow is Youtube with confidence "dns"
> 
> 
> If everything is implemented correctly we should got something like:
> 
> ```
> ./example/ndpiReader -t -i ~/fpc_dns.pcapng -v2
> 
> 
> 	1	TCP [2001:b07:a3d:c112:9326:58a1:68f7:5712]:40088 <-> [2a00:1450:4002:410::200e]:443 ... [FPC: 124/Youtube, Confidence: DNS]
> ```
> 
> Am I missing something?

@IvanNardi 

Thanks for detailed clarification,let me go through the attached dns packet capture and described logic flow with ongoing FPC implementation and update you any clarifications required.Thank you very much for quick support and all valuable advice.Really appreciate this advice to narrow down the implementation.

mmanoj commented:
> > > can you explain bit more about this? SRC_FLOW_IP + ip_required_in_the_dns_query is this flow source IP and DNS response IP (like A or AAA address?)
> > 
> > 
> > That was the idea, but I was wrong (because of ipv4 and ipv6 possible mismatch) We should use only one address as key.
> > Anyway, let try together to analyze what we should do step by step. In this simple trace there is a DNS flow and a TCP flow: [fpc_dns.zip](https://github.com/user-attachments/files/16239123/fpc_dns.zip)
> > ```
> > 1. We should take a look at every DNS A or AAAA replies. If the domain required provides a sub-classification, i.e. if `ndpi_match_host_subprotocol()` in DNS code returns a protocol different than `NDPI_PROTOCOL_UNKNOWN`, we should save that information into the cache. I think we can always add only one entry, even if there are multiple answers in the DNS reply
> >    *) Packet 2 is a DNS AAAA reply with Youtube as subclassification -> add an entry with key "2a00:1450:4002:416::200e" (i.e. the first answer) and protocol `NDPI_PROTOCOL_YOUTUBE`
> > 
> > 2. At the very first pkt of every flow, we should lookup into the cache using the server address (i.e. `flow->s_addr`), in `ndpi_internal_detection_process_packet` just after `fpc_check_ip()` call. If a match is found, we should call `fpc_update(ndpi_str, flow, NDPI_PROTOCOL_UNKNOWN, protocol_returned_from_the_cache, NDPI_FPC_CONFIDENCE_DNS)`
> >    *) At packet 3 we should lookup using "2a00:1450:4002:416::200e"  as key (the server address). The returned protocol is `NDPI_PROTOCOL_YOUTUBE`, so the FPC of this flow is Youtube with confidence "dns"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If everything is implemented correctly we should got something like:
> > ```
> > ./example/ndpiReader -t -i ~/fpc_dns.pcapng -v2
> > 
> > 
> > 	1	TCP [2001:b07:a3d:c112:9326:58a1:68f7:5712]:40088 <-> [2a00:1450:4002:410::200e]:443 ... [FPC: 124/Youtube, Confidence: DNS]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Am I missing something?
> 
> @IvanNardi
> 
> Thanks for detailed clarification,let me go through the attached dns packet capture and described logic flow with ongoing FPC implementation and update you any clarifications required.Thank you very much for quick support and all valuable advice.Really appreciate this advice to narrow down the implementation.

@IvanNardi 

The logic seems fine, let me implement and perform the initial test and update you for any clarifications required.

mmanoj commented:
> 2a00:1450:4002:416::200e
@IvanNardi 

I almost implement the logic from dns.c side and ndpi_main.c, still need to fix some the key generation logic with IP address as it's not working as expected.However hard coded key it's working fine.I will perform more test and share the code.

mmanoj commented:
> > 2a00:1450:4002:416::200e
> > @IvanNardi
> 
> I almost implement the logic from dns.c side and ndpi_main.c, still need to fix some the key generation logic with IP address as it's not working as expected.However hard coded key it's working fine.I will perform more test and share the code.

@IvanNardi 
Key generation issue fixed

	1	TCP 192.168.1.8:48516 <-> 157.240.235.60:443 [proto: 91.142/TLS.WhatsApp][IP: 142/WhatsApp][Encrypted][Confidence: DPI][**FPC: 242/WhatsAppFiles, Confidence: DNS**][DPI packets: 6][cat: Chat/9][179 pkts/21880 bytes <-> 178 pkts/155403 bytes][Goodput ratio: 46/92][12.09 sec][Hostname/SNI: web.whatsapp.com][(Advertised) ALPNs: h2;http/1.1][TLS Supported Versions: TLSv1.3;TLSv1.2][bytes ratio: -0.753 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 43/11 2752/306 264/38][Pkt Len c2s/s2c min/avg/max/stddev: 66/66 122/873 1103/1446 143/629][TLSv1.3][JA3C: a195b9c006fcb23ab9a2343b0871e362][JA4: t13d1715h2_5b57614c22b0_7121afd63204][JA3S: fcb2d4d0991292272fcb1e464eedfd43][ECH: version 0xfe0d][Firefox][Cipher: TLS_AES_128_GCM_SHA256][Plen Bins: 1,26,10,2,0,3,6,0,0,0,3,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,44,0,0,0,0]

mmanoj committed:
Merge branch 'ntop:dev' into dev

mmanoj committed:
FPC review change updates with DNS cache

FPC review change updates with DNS cache

mmanoj committed:
FPC review change updates with DNS cache

FPC review change updates with DNS cache

mmanoj committed:
FPC review change updates with DNS cache

FPC review change updates with DNS cache

mmanoj committed:
FPC review change updates with DNS cache

FPC review change updates with DNS cache

mmanoj commented:
@IvanNardi 

All changes updated with test pcap file also.Please review and advice if any changes required.

./ndpiReader -t -i fpc_dns.pcap -v2

1	TCP 192.168.1.8:39432 <-> 142.250.199.4:443 [proto: 91/TLS][IP: 126/Google][Encrypted][Confidence: DPI][**FPC: 126/Google, Confidence: DNS**][DPI packets: 3][cat: Web/5][405 pkts/47359 bytes <-> 742 pkts/962242 bytes][Goodput ratio: 44/95][4.54 sec][bytes ratio: -0.906 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 12/5 1094/1181 74/52][Pkt Len c2s/s2c min/avg/max/stddev: 66/66 117/1297 1466/1466 208/436][Risk: ** Probing attempt **][Risk Score: 50][Risk Info: TLS/QUIC Probing][Plen Bins: 3,3,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,91,0,0,0,0]

IvanNardi committed:
FPC (DNS): fix and improvements

Update documetation.
Fix key with ipv6 addreses over ipv4 flow.
Update unit tests results.

IvanNardi reviewed (approved):
I pushed some small changes and I'll update the commit message.
But the code was almost good anyway...
Thanks a lot for that and for your perseverance

mmanoj commented:
> I pushed some small changes and I'll update the commit message. But the code was almost good anyway... Thanks a lot for that and for your perseverance

@IvanNardi 

Thanks for the support and I was review the fixes you posted and it's good improvements which I was missed.I'm happy to do some valuable contribution to community and thank again for your advice and guidelines.Really appreciated.

Hope this change will improve the detection accuracy overall as per the test results I can see good detection specially services based on CDN's.

Once This pull request merge I will work on adding more IP lists from some reliable sources specially malware and C&C servers.
